src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1730,6 +1730,18 @@ bool checkValid(MethodSymbol candidate, bool isCandidate, DiagnosticBag specific
                         }
                     }
                 }
+                else if (LanguageVersion >= MessageID.IDS_FeatureAsyncMain.RequiredVersion() && taskEntryPoints.Count > 0)
```

Comment:
```

Expected:
                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("D.Main()"),
                Diagnostic(ErrorCode.WRN_SyncAndAsyncEntryPoints, "Main")
Actual:
                // (12,24): warning CS0028: 'D.Main()' has the wrong signature to be an entry point
                //   public static string Main() { System.Console.WriteLine(3); return null; }
                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("D.Main()").WithLocation(12, 24)
Diff:
                    Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("D.Main()")
-->                 Diagnostic(ErrorCode.WRN_SyncAndAsyncEntryPoints, "Main")
```

The warning wasn't produced. By looking to the if condition, I can't figure out what can cause the condition to be false.
The language version in the set is set to 7.1, to it's equal to FeatureAsyncMain.RequiredVersion, and there is a task entry point, so the Count is 1 (>0).

I can't debug such a large code on my poor machine. Sorry for that :-(

@RikkiGibson @333fred @gafter Can anyone help here?

Reply:
I have debugged this a bit. It does enter the if block on line 1733 and add the diagnostic successfully. Then the diagnostics get attached to the EntryPoint in `GetEntryPointAndDiagnostics`. Somewhere later the diagnostics seem to get dropped. I wasn't able to figure out why in the time I allotted to look at it.

Reply:
@RikkiGibson Thanks for taking the time. I found that entrypoint-related diagnostics are grouped in CSharpCompilerDiagnosticAnalyzer, I'm unsure of how things connect with each others, but I'll try to add the new warning in CSharpCompilerDiagnosticAnalyzer too and see if that fixes it.

Reply:
Where are you saying that this isn't produced? In your new test?

Reply:
@333fred, Yes the new test failed.

```
Expected:
                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("D.Main()"),
                Diagnostic(ErrorCode.WRN_SyncAndAsyncEntryPoints, "Main")
Actual:
                // (12,24): warning CS0028: 'D.Main()' has the wrong signature to be an entry point
                //   public static string Main() { System.Console.WriteLine(3); return null; }
                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("D.Main()").WithLocation(12, 24)
Diff:
                    Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("D.Main()")
-->                 Diagnostic(ErrorCode.WRN_SyncAndAsyncEntryPoints, "Main")
```

Reply:
If so, it's likely because the default warning level is still 4. You'll need to use `TestOptions.RegularExe.WithWarningLevel(5)` as the `options` parameter of `CreateCompilation`

Reply:
I've added WithWarningLevel(5). Hope that fixes it.